integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonShowSegmentsByQueryCommand.scala

```diff
@@ -0,0 +1,135 @@
+    setAuditTable(carbonTable)
+    if (!carbonTable.getTableInfo.isTransactionalTable) {
+      throw new MalformedCarbonCommandException("Unsupported operation on non transactional table")
+    }
+    try {
+      df.collect()
+    } catch {
+      case ex: Throwable =>
+        throw new MalformedCarbonCommandException("failed to run query: " + ex.getMessage)
+    } finally {
+      sparkSession.catalog.dropTempView(makeTempViewName(carbonTable))
+    }
+  }
+
+  override protected def opName: String = "SHOW SEGMENTS"
+
+  private def createDataFrame: DataFrame = {
+    val sparkSession = SparkSession.getActiveSession.get
+    val carbonTable = CarbonEnv.getCarbonTable(databaseNameOp, tableName)(sparkSession)
+    val tablePath = carbonTable.getTablePath
+    val segments = CarbonStore.readSegments(tablePath, showHistory)
+    val tempViewName = makeTempViewName(carbonTable)
+    registerSegmentRowView(sparkSession, tempViewName, carbonTable, segments)
+    try {
+      sparkSession.sql(query)
+    } catch {
+      case t: Throwable =>
+        sparkSession.catalog.dropTempView(tempViewName)
+        throw t
+    }
+  }
+
+  /**
+   * Generate temp view name for the query to execute
+   */
+  private def makeTempViewName(carbonTable: CarbonTable): String = {
+    s"${carbonTable.getTableName}_segments"
+  }
+
+  private def registerSegmentRowView(
+      sparkSession: SparkSession,
+      tempViewName: String,
+      carbonTable: CarbonTable,
+      segments: Array[LoadMetadataDetails]): Unit = {
+
+    // populate a dataframe containing all segment information
+    val tablePath = carbonTable.getTablePath
+    val segmentRows = segments.toSeq.map { segment =>
+      val mergedToId = CarbonStore.getMergeTo(segment)
+      val path = CarbonStore.getSegmentPath(segment)
+      val startTime = CarbonStore.getLoadStartTime(segment)
+      val endTime = CarbonStore.getLoadEndTime(segment)
+      val spentTime = CarbonStore.getSpentTimeAsMillis(segment)
+      val (dataSize, indexSize) = CarbonStore.getDataAndIndexSize(tablePath, segment)
+      val partitions = CarbonStore.getPartitions(tablePath, segment)
+      SegmentRow(
+        segment.getLoadName,
+        segment.getSegmentStatus.toString,
+        startTime,
+        spentTime,
+        partitions,
+        dataSize,
+        indexSize,
+        mergedToId,
+        segment.getFileFormat.toString,
+        path,
+        endTime,
+        if (segment.getSegmentFile == null) "NA" else segment.getSegmentFile)
+    }
+
+    // create a temp view using the populated dataframe and execute the query on it
+    val df = sparkSession.createDataFrame(segmentRows)
+    checkIfTableExist(sparkSession, tempViewName)
+    df.createOrReplaceTempView(tempViewName)
+  }
+
+  private def checkIfTableExist(sparkSession: SparkSession, tempViewName: String): Unit = {
+    if (sparkSession.catalog.tableExists(tempViewName)) {
+      throw new MalformedCarbonCommandException(s"$tempViewName already exists, " +
```

Comment:
i have a suggestion, instead of failing the shoe segments by query, just because he cannot create temp view, if temp view already exists can we can fallback to basicInfo, So atleast user will get info and query part he can try later. This will help in concurrent scenario. What do you think?

Reply:
I think it is better to fail it since user may depends on the result to do something programatically. If we return something not expected, then the behavior is hard to understand